hail/python/hail/utils/misc.py

```diff
@@ -291,58 +291,125 @@ def get_nice_field_error(obj, item):
             from hail.expr.expressions import ExpressionException
-            raise ExpressionException("selection would produce duplicate field '{}'".format(repr(k)))
+            raise ExpressionException(f"{caller!r}: selection would produce duplicate field {k!r}")
+
+def get_key_by_exprs(caller, exprs, named_exprs, indices, override_protected_indices=None):
+    from hail.expr.expressions import to_expr, ExpressionException, analyze
+    exprs = [indices.source[e] if isinstance(e, str) else e for e in exprs]
+    named_exprs = {k: to_expr(v) for k, v in named_exprs.items()}
+
+    bindings = []
+
+    def is_top_level_field(e):
+        return e in indices.source._fields_inverse
+
+    existing_key_fields = []
+    final_key = []
+    for e in exprs:
+        analyze(caller, e, indices, broadcast=False)
+        if not e._ir.is_nested_field:
+            raise ExpressionException(f"{caller!r} expects keyword arguments for complex expressions\n"
+                                      f"  Correct:   ht = ht.key_by('x')\n"
+                                      f"  Correct:   ht = ht.key_by(ht.x)\n"
+                                      f"  Correct:   ht = ht.key_by(x = ht.x.replace(' ', '_'))\n"
+                                      f"  INCORRECT: ht = ht.key_by(ht.x.replace(' ', '_'))")
+
+        name = e._ir.name
+        final_key.append(name)
+
+        if not is_top_level_field(e):
+            bindings.append((name, e))
+        else:
+            existing_key_fields.append(name)
+
+    final_key.extend(named_exprs)
+    bindings.extend(named_exprs.items())
+    check_collisions(caller, final_key, indices, override_protected_indices=override_protected_indices)
+    return final_key, dict(bindings)
 
-def check_keys(name, indices):
+
+def check_keys(caller, name, protected_key):
     from hail.expr.expressions import ExpressionException
-    if indices.key is None:
-        return
-    if name in set(indices.key):
-        msg = "cannot overwrite key field {} with annotate, select or drop; use key_by to modify keys.".format(repr(name))
+    if name in protected_key:
+        msg = f"{caller!r}: cannot overwrite key field {name!r} with annotate, select or drop; " \
+              f"use key_by to modify keys."
         error('Analysis exception: {}'.format(msg))
         raise ExpressionException(msg)
 
-def get_select_exprs(caller, exprs, named_exprs, indices, protect_keys=True):
+def get_select_exprs(caller, exprs, named_exprs, indices, base_struct):
     from hail.expr.expressions import to_expr, ExpressionException, analyze
-    exprs = [to_expr(e) if not isinstance(e, str) else indices.source[e] for e in exprs]
+    exprs = [indices.source[e] if isinstance(e, str) else e for e in exprs]
     named_exprs = {k: to_expr(v) for k, v in named_exprs.items()}
-    assignments = OrderedDict()
+    select_fields = indices.protected_key[:]
+    protected_key = set(select_fields)
+    insertions = []
+
+    final_fields = select_fields[:]
```

Comment:
two copies, yes. Maybe Indices.protected_key should return a tuple or something (Which would **need** to be copied), but the set copy is for faster membership checks, and the copy on line 367 is because we intend to mutate `final_fields`.

Reply:
three copies, `select_fields` and `protected_key` are both copies as well, but it looks like you need all three copies.